beacon-chain/sync/backfill/blobs.go

```diff
@@ -0,0 +1,134 @@
+	store    das.AvailabilityStore
+	expected []blobSummary
+	next     int
+	bbv      *blobBatchVerifier
+	cs       primitives.Slot
+}
+
+func (bs *blobSync) blobsNeeded() int {
+	return len(bs.expected) - bs.next
+}
+
+func (bs *blobSync) validateNext(rb blocks.ROBlob) error {
+	if bs.next >= len(bs.expected) {
+		return errUnexpectedResponseSize
+	}
+	next := bs.expected[bs.next]
+	bs.next += 1
+	// Get the super cheap verifications out of the way before we init a verifier.
+	if next.blockRoot != rb.BlockRoot() {
+		return errors.Wrapf(errUnexpectedResponseContent, "next expected root=%#x, saw=%#x", next.blockRoot, rb.BlockRoot())
+	}
+	if next.index != rb.Index {
+		return errors.Wrapf(errUnexpectedResponseContent, "next expected root=%#x, saw=%#x for root=%#x", next.index, rb.Index, next.blockRoot)
+	}
+	if next.commitment != bytesutil.ToBytes48(rb.KzgCommitment) {
+		return errors.Wrapf(errUnexpectedResponseContent, "next expected commitment=%#x, saw=%#x for root=%#x", next.commitment, rb.KzgCommitment, rb.BlockRoot())
+	}
+
+	if bytesutil.ToBytes96(rb.SignedBlockHeader.Signature) != next.signature {
+		return verification.ErrInvalidProposerSignature
+	}
+	v := bs.bbv.newVerifier(rb)
+	if err := v.BlobIndexInBounds(); err != nil {
+		return err
+	}
+	v.SatisfyRequirement(verification.RequireValidProposerSignature)
+	if err := v.SidecarInclusionProven(); err != nil {
+		return err
+	}
+	if err := v.SidecarKzgProofVerified(); err != nil {
+		return err
+	}
+	if err := bs.store.Persist(bs.cs, rb); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func newBlobBatchVerifier(nbv verification.NewBlobVerifier) *blobBatchVerifier {
+	return &blobBatchVerifier{newBlobVerifier: nbv, verifiers: make(blobVerifierMap)}
+}
+
+type blobBatchVerifier struct {
+	newBlobVerifier verification.NewBlobVerifier
+	verifiers       blobVerifierMap
+}
+
+func (bbv *blobBatchVerifier) newVerifier(rb blocks.ROBlob) verification.BlobVerifier {
+	m := bbv.verifiers[rb.BlockRoot()]
+	m[rb.Index] = bbv.newBlobVerifier(rb, verification.BackfillSidecarRequirements)
+	bbv.verifiers[rb.BlockRoot()] = m
+	return m[rb.Index]
+}
+
+func (bbv blobBatchVerifier) VerifiedROBlobs(_ context.Context, blk blocks.ROBlock, _ []blocks.ROBlob) ([]blocks.VerifiedROBlob, error) {
+	m, ok := bbv.verifiers[blk.Root()]
+	if !ok {
+		return nil, errors.Wrapf(verification.ErrMissingVerification, "no record of verifiers for root %#x", blk.Root())
+	}
+	c, err := blk.Block().Body().BlobKzgCommitments()
+	if err != nil {
+		return nil, errors.Wrapf(errUnexpectedCommitment, "error reading commitments from block root %#x", blk.Root())
+	}
+	vbs := make([]blocks.VerifiedROBlob, len(c))
+	for i := range c {
+		if m[i] == nil {
+			return nil, errors.Wrapf(errBatchVerifierMismatch, "do not have verifier for block root %#x idx %d", blk.Root(), i)
+		}
+		vb, err := m[i].VerifiedROBlob()
+		if err != nil {
+			return nil, err
+		}
+		if bytesutil.ToBytes48(vb.KzgCommitment) != bytesutil.ToBytes48(c[i]) {
```

Comment:
q: is this better than `!bytes.Equal`? I think ToBytes48 allocates memory while bytes.Equal casts to string to compare in place

Reply:
maybe we can also just do `[48]byte(vb.KzgCommitment) != [48]byte(c[i])` .

Reply:
updated